### ISSUE TITLE
Progress bar doesn't respect borders of tmux pane

### ISSUE THREAD

**Issue by kalekseev** [kennethreitz/pipenv#981]
Progress bar goes out of tmux pane border

##### Describe you environment

1. Mac OS high sierra
1. Python version: 3.6.2
1. Pipenv version: 8.2.7

##### Expected result

Progress bar works as in the real terminal

##### Actual result

![2017-10-27 14 38 26](https://user-images.githubusercontent.com/367259/32102502-4b9648ca-bb25-11e7-8dda-bd4fdb2a747e.png)

##### Steps to replicate

Run tmux, split vertically, go to the right pane, run `pipenv install`

**Comment by piotr-dobrogost** [comment]
Out of curiosity; what is the function of Apple's logo in prompt?

**Comment by kalekseev** [comment]
@piotr-dobrogost I believe it's information about file format (like unix), don't care much about that so can't say for sure, I think if you install vim-airline and vim-devicons then this works automatically.

![2017-10-27 16 20 59](https://user-images.githubusercontent.com/367259/32105987-19b200fc-bb33-11e7-8ce6-d1f04705dd12.png)

**Comment by techalchemy** [comment]
I’m guessing this relates to how we are retrieving terminal size info. It seems reasonably likely this would be a bug, although I never vsplit my tmux session.

**Comment by erinxocon** [comment]
@kalekseev Out of curiousity, what happens if you resize the tmux split pane after you create it.  We have code in there to adjust layouts when the terminal is resized and I'm wondering if the split isn't considered a resize.

**Comment by techalchemy** [comment]
I actually didn't have this problem -- I split my tmux session, jumped to the right hand side, and put enough crap in my Pipfile to get a progress bar.  Based on your progress bar this looks like an old version of pipenv.  @kalekseev can you try upgrading pipenv and test again?

![image](https://user-images.githubusercontent.com/954310/32147473-32392fd8-bcbe-11e7-9eec-0cc396202844.png)

**Comment by erinxocon** [comment]
I can confirm that the issue seems to be non existent to me.  Closing issue due to probable old version!  Let us know if that's not the case @kalekseev

**Comment by kalekseev** [comment]
@erinxocon  upgraded to pipenv 8.3.1, removed custom tmux.conf the problem still exists in iterm 3.1.4, though it works as expected in the Terminal.app
![2017-10-30 11 29 56](https://user-images.githubusercontent.com/367259/32161192-b45ab20a-bd65-11e7-84ab-53454f94ff87.png)

**Comment by erinxocon** [comment]
@kalekseev  huh, I was using iterm as well.  Different config?  That's weird.

**Comment by techalchemy** [comment]
Thanks for the info. This isn’t super high priority and supporting this kind of progress bar on multiplexers may not even be realistic but it’s not really my area of expertise. Would be interested in feedback

**Comment by davesque** [comment]
Yeah, not a very serious bug.  Although there could be a hidden cost if it affects people's perception of the tech and, therefore, the adoption rate.  For example, I've never experienced similar terminal-breaking behavior with the package managers from other languages/platforms.  If I were some kind of manager less familiar with Python and I was making a call as to whether or not my department should adopt it, little dumb things like that might have a subtle impact on my decision.  I had a conversation just the other day with a department manager who seemed to believe, for some reason, that Python's package management solutions were not up to par compared with other languages.  I didn't agree but, somehow, this person had come to hold this opinion.

**Comment by uranusjr** [comment]
@davesque Out of curiosity, do you know what the manager considers good package management solutions? Python’s packaging has been drastically improved recently to become acceptable, but I understand people still maintaining it being bad if they don’t stay on the cutting edge. The situation is still being discussed a lot and improved upon, and any advices would be welcomed. It’d be interesting to hear what people thinks pip-related tooling can learn from other package managers.

**Comment by davesque** [comment]
@uranusjr 

I think it was a node.js/Ruby developer I was speaking to although I don't think I could say with any certainty what motivated that person to hold that position.  My loose impression that it was something like what you say---that this person formed their opinion a while back when there really was more friction in Python's package management process but is now holding on to it just because it was a first impression.

Again, I don't share their views on the matter since I have been using Python and all of its tooling for a while now and I've become aware of the professionalism that goes into the design of all the resources used by the community.

Anyway, my point was just that little polish issues like this can have a surprising amount of importance when you consider them from a certain point of view.

**Comment by techalchemy** [comment]
It’d be helpful if we can narrow this down: is this Mac only? Iterm only? Locale setting specific? What configurations are in play for fonts, terminal, tmux etc? Are all of the things up to date?

**Comment by uranusjr** [comment]
I played around a bit and this seems to have something to do with how iTerm2 reports Unicode character widths. Setting “Use Unicode Version 9 Widths” in Preferences → Profiles → (The profile you use) → Text fixes it for me.

[iTerm2’s documentation](https://www.iterm2.com/documentation-one-page.html) seems to suggest you should set it unless you have faulty legacy application that malfunctions. New versions of Tmux built with recommended configurations use the new width table.

So to summarise:

* Check the aforementioned checkbox.
* Consider filing an issue to iTerm2 to make this the default.
* If for whatever reason you can’t use the up-to-date width table, set environment variable `PIPENV_HIDE_EMOJIS` to workaround the issue.

**Comment by techalchemy** [comment]
Is there any way we can tell if the checkbox is checked?

**Comment by uranusjr** [comment]
You can tell you’re in iTerm2 with `TERM_PROGRAM`, Tmux with `TMUX` (or `TERM` if you want to consider GNU Screen as well, which could also be affected, though I didn’t try), and iTerm preferences with [`defaults`](https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man1/defaults.1.html). So yes, you can. I don’t think it would be appropriate in practice though.

Also you can’t actually tell what mode the terminal is in, only guess. The output is identical from the program’s perspective; it’s iTerm *rendering* things differently.

**Comment by davesque** [comment]
@uranusjr Thanks for the insight into this.  Created an issue in iTerm's gitlab repo here: https://gitlab.com/gnachman/iterm2/issues/6601

**Comment by gnachman** [comment]
iTerm2 defaults you to unicode 9 if you're on macOS 10.13, but if you've ever fiddled with the setting before it'll stay at what you set it to.

**Comment by davesque** [comment]
@gnachman I'll assume that's why this happened.  I've been a user of iTerm for a while now so I probably had an old config sitting around.
